#1295 Write WMH count to the subject dir

jan-petr opened the pull request:
### Linked issue

Closes #1295

jan-petr committed:
#1295 Write WMH count to the subject dir

HenkMutsaerts commented on External/SPMmodified/xASL/xASL_bids_csv2tsvReadWrite.m:
```diff
@@ -1,7 +1,7 @@
-function [PathTSV, CellContents] = xASL_bids_csv2tsvReadWrite(PathIn, bDeleteCSV, bWriteTSV)
+function [PathTSV, CellContents] = xASL_bids_csv2tsvReadWrite(PathIn, bDeleteCSV, bOverwrite)
```
All changes of `bWriteTSV` to `bOverwrite` are illegal here. You can add `bOverwrite`, but you cannot remove `bWriteTSV`.

If `bWriteTSV` is false, and you read a CSV, it reads the content and puts it in memory. If `bWriteTSV` is true, it also reads to memory but additionally converts to TSV.

jan-petr replied:
Not entirely true - the code previously had bOverwrite in the code description, so I was only cleaning code. Not adding anything. So up to you to decide what names you want. I was just syncing it...

HenkMutsaerts replied:
The functionality is very different. `bOverwrite` means just overwriting yes or no. `bWriteTSV` means if you should convert a CSV-file to a TSV file (or only output the read CSV-file to memory). So this needs reverting.

jan-petr replied:
Agree. I was confused and tried to clean it. But I agree with your change.

HenkMutsaerts commented on External/SPMmodified/xASL/xASL_bids_csv2tsvReadWrite.m:
```diff
@@ -45,7 +45,7 @@
 [Fpath, Ffile, Fext] = xASL_fileparts(PathIn);
 
-if isempty(regexp(Fext, '^\.(tsv|csv)$'))
+if isempty(regexp(Fext, '^\.(tsv|csv)$', 'once'))
```
You keep adding 'once' because Matlab complaints, I would argue this just makes stuff less readable, we won't notice the speed increase ;)

jan-petr replied:
The speed is not important - but it removes Matlab warnings - which is nice to have. Not to optimize code, but to have code without warnings. But we can for sure skip this the next time...

HenkMutsaerts commented on Modules/SubModule_Structural/xASL_wrp_GetVolumetrics.m:
```diff
@@ -94,21 +94,29 @@ function xASL_wrp_GetVolumetrics(x)
     
-    % run the LST lesion volume calculator
+    % Run the LST lesion volume calculator
+	% Change directory to the SUBJECT DIRECTORY as the count function outputs to the current directory
```
"count function" is not clear, it took me some time to understand that you meant `ps_LST_tlv`, (am I right?) Please be explicit ;)

jan-petr replied:
sure - you can update the comment...

HenkMutsaerts replied:
Done

HenkMutsaerts commented on Modules/SubModule_Structural/xASL_wrp_GetVolumetrics.m:
```diff
@@ -94,21 +94,29 @@ function xASL_wrp_GetVolumetrics(x)
     % define new filename:
-    OutputPath = fullfile(x.D.TissueVolumeDir, ['WMH_LST_' x.WMHsegmAlg '_' x.P.SubjectID '.csv']);
-    % Then move our file
+    OutputPathCSV = fullfile(x.dir.SUBJECTDIR, ['WMH_LST_' x.WMHsegmAlg '_' x.P.SubjectID '.csv']);
```
From here below you need to revert, it is too long and wrong; this doesn't concern temporary files, and the output should remain in the population folder.

We can eventually move all structural output to subject folders, but that would be a large future issue.

jan-petr replied:
Actually. It has to stay like this. The temporary files in population folder were completely messed up during parallelization. As this is structural module outputting to population folder. If you still want to output temp files there, then you need to change the entire structure and debug. I don't see issue outputting temp files to subject folder.

HenkMutsaerts replied:
No, I agreed with the first part where you change the temporary file to the subject folder (assuming LST outputs the same name for each subject, which gets messed up during parallelization).

It was just the last part where you added unnecessary code.

jan-petr replied:
OK. Now I understand! But there's still an issue.
1. The former version saved the temporary to POPULATION dir and then renamed it - this was causing issues in parallelization.
2. So I made a new version that saved the temporary to SUBJECTDIR and then moved the final to POPULATION. But I agree that it was way to complicated.
3. But you edit saves everything to SUBJECTDIR and never moves it to the population.
4. So I made a last fix that saves temporary to SUBJECTDIR and then moves it to POPULATION and there converts to TSV. Please check if this is clean enough.

HenkMutsaerts reviewed (changes requested):
How much time did you spend on this, would it be worth checking with me first?

jan-petr commented:
> How much time did you spend on this, would it be worth checking with me first?

Not much. But read again - this has to be fixed. The previous implementation was causing bugs in parallel mode...

HenkMutsaerts committed:
#1295 xASL_wrp_GetVolumetrics: revert code complication

HenkMutsaerts committed:
#1295 xASL_wrp_GetVolumetrics: made comment readable;)

HenkMutsaerts committed:
#1295 xASL_bids_csv2tsvReadWrite: bOverwrite->bWriteTSV

Also improved the header so it is clear what this parameters does, both converting a CSV to TSV & overwriting any pre-existing TSV

HenkMutsaerts reviewed (approved):
See my comments, this should fix my issues

jan-petr committed:
#1295 Move WMH count file to the population dir

HenkMutsaerts commented on External/SPMmodified/xASL/xASL_bids_csv2tsvReadWrite.m:
```diff
@@ -64,6 +64,15 @@
     end
 end
 
+if strcmp(Fext, '.tsv') && ~exist(PathIn, 'file')
```
This was not what this function was intended to. This function was to read a CSV-file and at the same time convert to a TSV-file, which happens a lot when taking output of LST, CAT12, etc, and converting to BIDS (BIDS likes TSV).

Otherwise, we would simply use `xASL_tsvRead` or `xASL_csvRead`.

Was this not clear from the header?

I'm not sure your change here changes the concept, because if we already have a TSV-file, LST outputs a CSV-file with the same name, it should always overwrite. Now, it would take the old TSV-context and save it as if this was the new TSV.

So we should probably revert your edit here, but the header should be more clear. And rename this function, so it is more clear, e.g. `xASL_bids_csv2tsv.m` instead of ReadWrite

jan-petr replied:
I don't want to do "a Henk" - but please read my comments before you reply :)
I know what was this issue supposed to do - previously, it saved the CSV in a way that it was conflicting and crashing during parallelization, so we had to save it elsewhere, convert CSV to TSV and then move TSV to the population directory. See all my comments above. I'm fine with the csv2tsv function but it has to be run in the subject dir...

jan-petr replied:
We can remove this commit, but I just wanted to catch all possible misuses of this function. So this is a safety net, not a function ;)But we can remove it if you want...

HenkMutsaerts commented on Modules/SubModule_Structural/xASL_wrp_GetVolumetrics.m:
```diff
@@ -95,22 +95,26 @@ function xASL_wrp_GetVolumetrics(x)
     if ~exist(FileName, 'file')
          warning(['Missing:' FileName]);
 	else
+		% First rename the output file to the correct filename format in the population dir
```
If you remove the word "format" here, I'm fine :)
Because the format is changed below.

And I like that you make the comments this explicit! ;)

jan-petr replied:
that was to convince you :)
Sure "format" was a typo ;)

jan-petr replied:
I'll do this once I'm done with running other tests. So please only check the function above.

HenkMutsaerts reviewed (changes requested):
Minor question but mostly fine
